### PR TITLE
Make bootstrapping easier by avoiding custom Setup.hs

### PR DESCRIPTION
**Pull request by phadej**
Resolves https://github.com/simonmar/happy/issues/144

### CODE REVIEW

**Comment by phadej** [comment]
Travis is turning green as well.

File: gen-happy-sdist/gen-happy-sdist.cabal [hunk 1]
```diff
@@ -0,0 +1,15 @@
+cabal-version:       1.12
+name:                gen-alex-sdist
```
Comment (simonmar) [inline gen-happy-sdist/gen-happy-sdist.cabal]: copy/paste-o

File: gen-happy-sdist/gen-happy-sdist.cabal [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+version:             0
+synopsis:            Tool for generating `alex` sdist
```
Comment (simonmar) [inline gen-happy-sdist/gen-happy-sdist.cabal]: ditto

File: gen-happy-sdist/gen-happy-sdist.cabal [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+category:            Development
+build-type:          Simple
+
+executable gen-alex-sdist
```
Comment (simonmar) [inline gen-happy-sdist/gen-happy-sdist.cabal]: ditto

File: happy.cabal [hunk 4]
```diff
@@ -177,5 +192,8 @@ executable happy
 test-suite tests
   type: exitcode-stdio-1.0
   main-is: test.hs
+  -- This line is important as it ensures that the local `exe:alex` component declared above is built before the test-suite component is invoked, as well as making sure that `alex` is made available on $PATH and `$alex_datadir` is set accordingly before invoking `test.hs`
```
Comment (simonmar) [inline happy.cabal]: s/alex/happy?

File: gen-happy-sdist/gen-happy-sdist.cabal [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  default-language:    Haskell2010
+  main-is:             Main.hs
+  build-depends:       base                    < 5
+                     , cpphs      >= 1.20.8 && < 1.21
```
Comment (simonmar) [inline gen-happy-sdist/gen-happy-sdist.cabal]: Is cpphs needed? Why?

Comment (phadej) [reply gen-happy-sdist/gen-happy-sdist.cabal]: Templates are processed with CPP. `cpphs` is safe choice for CPP, than simply relying on whatever version one might have on their machine, i.e. we'll produce deterministic results. Also easy to use as a library.

Yet, I really don't know. `gen-alex-sdist` uses `cpphs` as well: https://github.com/simonmar/alex/blob/1e7af951077eb39ce3a0b669fbbb64ae4e3f62b2/gen-alex-sdist/gen-alex-sdist.cabal#L12

**Comment by phadej** [comment]
The job failure seems to be flaky network. I cannot restart it myself.

**Comment by hvr** [comment]
@phadej just force-push an amended commit to this git branch/pr

File: gen-happy-sdist/gen-happy-sdist.cabal (hunk 5 above)
Comment (hvr) [reply gen-happy-sdist/gen-happy-sdist.cabal]: indeed; Oleg's rationale is basically why I used cpphs so we'd have deterministic OS-independent results (and wouldn't have to deal with compat issues between various `cpp` implementations which have subtly different corner-case behaviour)

Comment (simonmar) [reply gen-happy-sdist/gen-happy-sdist.cabal]: OK

**Pull request merged**